File: api/src/main/java/org/pmiops/workbench/notebooks/NotebooksRetryHandler.java
package org.pmiops.workbench.notebooks;

import jakarta.inject.Provider;
import jakarta.servlet.http.HttpServletResponse;
import java.net.SocketTimeoutException;
import java.util.logging.Logger;
import org.broadinstitute.dsde.workbench.client.sam.api.TermsOfServiceApi;
import org.pmiops.workbench.exceptions.ExceptionUtils;
import org.pmiops.workbench.exceptions.WorkbenchException;
import org.pmiops.workbench.terra.TerraServiceRetryHandler;
import org.pmiops.workbench.utils.ResponseCodeRetryPolicy;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.retry.backoff.BackOffPolicy;
import org.springframework.stereotype.Service;

@Service
public class NotebooksRetryHandler extends TerraServiceRetryHandler<ApiException> {

  private static final Logger logger = Logger.getLogger(NotebooksRetryHandler.class.getName());

  private static class NotebookRetryPolicy extends ResponseCodeRetryPolicy {

    public NotebookRetryPolicy() {
      super("Notebooks API");
    }

    @Override
    protected int getResponseCode(Throwable lastException) {
      if (lastException instanceof ApiException) {
        return ((ApiException) lastException).getCode();
      }
      if (lastException instanceof SocketTimeoutException) {
        return HttpServletResponse.SC_GATEWAY_TIMEOUT;
      }
      return HttpServletResponse.SC_INTERNAL_SERVER_ERROR;
    }

    @Override
    protected void logNoRetry(Throwable t, int responseCode) {
      if (t instanceof ApiException) {
        logger.log(
            getLogLevel(responseCode),
            String.format(
                "Exception calling Notebooks API with response: %s",
                ((ApiException) t).getResponseBody()),
            t);
      } else {
        super.logNoRetry(t, responseCode);
      }
    }
  }

  @Autowired
  public NotebooksRetryHandler(
      BackOffPolicy backoffPolicy, Provider<TermsOfServiceApi> termsOfServiceApiProvider) {
    super(
        backoffPolicy,
        new NotebookRetryPolicy(),
        termsOfServiceApiProvider,
        ExceptionUtils::convertNotebookException);
  }

  @Override
  protected WorkbenchException convertException(ApiException exception) {
    return convertTerraException(exception, exception.getCode());
  }
}

File: api/src/main/java/org/pmiops/workbench/notebooks/NotebooksRetryHandler.kt
package org.pmiops.workbench.notebooks

import java.net.SocketTimeoutException
import java.util.logging.Logger
import javax.servlet.http.HttpServletResponse
import org.pmiops.workbench.config.RetryConfig
import org.pmiops.workbench.exceptions.ExceptionUtils
import org.pmiops.workbench.exceptions.WorkbenchException
import org.pmiops.workbench.utils.RetryHandler
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.retry.backoff.BackOffPolicy
import org.springframework.stereotype.Service

@Service
class NotebooksRetryHandler @Autowired
constructor(backoffPolicy: BackOffPolicy) : RetryHandler<ApiException>(backoffPolicy, NotebookRetryPolicy()) {

    private class NotebookRetryPolicy : RetryConfig.ResponseCodeRetryPolicy("Notebooks API") {

        override fun getResponseCode(lastException: Throwable): Int {
            if (lastException is ApiException) {
                return (lastException as ApiException).getCode()
            }
            return if (lastException is SocketTimeoutException) {
                HttpServletResponse.SC_GATEWAY_TIMEOUT
            } else HttpServletResponse.SC_INTERNAL_SERVER_ERROR
        }

        override fun logNoRetry(t: Throwable, responseCode: Int) {
            if (t is ApiException) {
                logger.log(
                        getLogLevel(responseCode),
                        String.format(
                                "Exception calling Notebooks API with response: %s",
                                (t as ApiException).getResponseBody()),
                        t)
            } else {
                super.logNoRetry(t, responseCode)
            }
        }
    }

    override fun convertException(exception: ApiException): WorkbenchException {
        return ExceptionUtils.convertNotebookException(exception)
    }

    companion object {

        private val logger = Logger.getLogger(NotebooksRetryHandler::class.java.name)
    }
}
